File: 2019/day-16.js
var fs = require('fs');
var text = fs.readFileSync('day-16.txt', 'utf8');
// var text = '19617804207202209144916044189917';
var bigArray = text.split('');

var counter = bigArray.length;

const phase = '010-1';
const PHASED = {
	A: 0,
	B: 1,
	C: -1
};

function createPhase(n) {
	let res = [];

	while (res.length < counter+1) {
		res.push('A'.repeat(n));
		res.push('B'.repeat(n));
		res.push('A'.repeat(n));
		res.push('C'.repeat(n));
	}

	res = res.join('');
	res = res.substr(1);
	res = res.split('').map(el => PHASED[el]);

	return res;
}

let INPUT = bigArray;

function doPhase() {
	let res;
	let NEW_INPUT = INPUT;
	for(var i = 0; i < counter; i++) {
		let phase = createPhase(i+1);
		res = INPUT.map((el, index) => {
			return Number(el) * Number(phase[index])
		});

		let sum = res.reduce((el, sum) => el + sum, 0);
		NEW_INPUT[i] = sum.toString().substr(-1);
	}

	return NEW_INPUT;
}

function star() {
	for(var i = 0; i < 100; i++) {
		start = doPhase();
	}

	// return start.join('');
	return start.slice(0,8).join('');
}

console.log(star());

File: 2019/day-16.py
import math

file = open('day-16.txt');
dest = file.read();
arr = list(dest);
phase = '010-1';

PHASED = {
	'A': 0,
	'B': 1,
	'C': -1
};

INPUT = arr;

def createPhase(n):
	res = [];
	counter = len(INPUT) + 1;
	while len(res) < counter:
		res.extend('A'*n);
		res.extend('B'*n);
		res.extend('A'*n);
		res.extend('C'*n);

	res = ''.join(res);
	res = res[1:];
	res = [PHASED[el] for el in res];

	return res;
		

def doPhase():
	new_input = INPUT;
	counter = len(INPUT);
	for i in range(counter):
		phase = createPhase(i+1);
		res = [int(el) * phase[idx] for idx, el in enumerate(INPUT)];

		res_sum = sum(res);
		new_input[i] = str(res_sum)[-1];

	return new_input;

def star():
	res = 0;
	for i in range(100):
		res = doPhase();

	print(''.join(res[:8]))
	return res;

star()

